In [2]:
import requests

In [3]:
import pandas as pd

In [4]:
url = "https://jsonplaceholder.typicode.com/comments"

In [5]:
response = requests.get(url)

In [6]:
data = response.json()

In [7]:
print(len(data))

500


In [8]:
df = pd.DataFrame(data)

In [9]:
print(df.head())

   postId  id                                       name  \
0       1   1               id labore ex et quam laborum   
1       1   2  quo vero reiciendis velit similique earum   
2       1   3              odio adipisci rerum aut animi   
3       1   4                             alias odio sit   
4       1   5      vero eaque aliquid doloribus et culpa   

                    email                                               body  
0      Eliseo@gardner.biz  laudantium enim quasi est quidem magnam volupt...  
1  Jayne_Kuhic@sydney.com  est natus enim nihil est dolore omnis voluptat...  
2     Nikita@garfield.biz  quia molestiae reprehenderit quasi aspernatur\...  
3           Lew@alysha.tv  non et atque\noccaecati deserunt quas accusant...  
4       Hayden@althea.biz  harum non quasi et ratione\ntempore iure ex vo...  


In [10]:
print(df.shape)

(500, 5)


In [11]:
print(df.columns)

Index(['postId', 'id', 'name', 'email', 'body'], dtype='str')


In [12]:
print(df.dtypes)

postId    int64
id        int64
name        str
email       str
body        str
dtype: object


In [13]:
df.to_csv('jsonplaceholder_comments.csv', index=False)

### PART 2 DATA CLEANING

In [15]:
clean_df =pd.read_csv("jsonplaceholder_comments.csv")

In [16]:
print(clean_df.head())

   postId  id                                       name  \
0       1   1               id labore ex et quam laborum   
1       1   2  quo vero reiciendis velit similique earum   
2       1   3              odio adipisci rerum aut animi   
3       1   4                             alias odio sit   
4       1   5      vero eaque aliquid doloribus et culpa   

                    email                                               body  
0      Eliseo@gardner.biz  laudantium enim quasi est quidem magnam volupt...  
1  Jayne_Kuhic@sydney.com  est natus enim nihil est dolore omnis voluptat...  
2     Nikita@garfield.biz  quia molestiae reprehenderit quasi aspernatur\...  
3           Lew@alysha.tv  non et atque\noccaecati deserunt quas accusant...  
4       Hayden@althea.biz  harum non quasi et ratione\ntempore iure ex vo...  


In [17]:
print(clean_df.info())

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   postId  500 non-null    int64
 1   id      500 non-null    int64
 2   name    500 non-null    str  
 3   email   500 non-null    str  
 4   body    500 non-null    str  
dtypes: int64(2), str(3)
memory usage: 19.7 KB
None


In [18]:
print(clean_df.describe())

           postId          id
count  500.000000  500.000000
mean    50.500000  250.500000
std     28.894979  144.481833
min      1.000000    1.000000
25%     25.750000  125.750000
50%     50.500000  250.500000
75%     75.250000  375.250000
max    100.000000  500.000000


In [19]:
print(clean_df.isnull().sum())

postId    0
id        0
name      0
email     0
body      0
dtype: int64


In [20]:
#To get missing values in the DataFrame, we can use the isnull() method along with sum() to count the number of missing values in each column.
missing_count = clean_df.isnull().sum()

In [21]:
missing_piece = (missing_count/len(clean_df))*100

In [22]:
missing_summary = pd.DataFrame({'Missing Count': missing_count, 'Missing Percentage': missing_piece})

In [23]:
print(missing_summary)

        Missing Count  Missing Percentage
postId              0                 0.0
id                  0                 0.0
name                0                 0.0
email               0                 0.0
body                0                 0.0


In [24]:
clean_df.isnull().sum()

postId    0
id        0
name      0
email     0
body      0
dtype: int64

In [25]:
#No missing values found.
(missing_count / len(clean_df)) * 100

postId    0.0
id        0.0
name      0.0
email     0.0
body      0.0
dtype: float64

In [26]:
duplicate_count = clean_df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


In [27]:
print(clean_df[clean_df.duplicated()])

Empty DataFrame
Columns: [postId, id, name, email, body]
Index: []


In [28]:
#Get the data types present in the DataFrame using the dtypes attribute. This will show the data type of each column in the DataFrame.Frame
print(clean_df.dtypes)

postId    int64
id        int64
name        str
email       str
body        str
dtype: object


In [30]:
### Data Type Decision

#The numeric columns (`userId` and `id`) were already stored as integer values, so no conversion was necessary. The `title` and `body` columns were correctly stored as text. The dataset does not contain a date column, so no date conversion was required.

In [31]:
df["id"].max()

np.int64(500)

In [32]:
df["id"].min()

np.int64(1)

In [34]:
cleaned_df = df.copy()

In [36]:
cleaned_df = df.drop_duplicates()

In [38]:
cleaned_df.shape

(500, 5)

## Cleaning Decisions

- Missing values were checked using `.isnull().sum()` and percentage calculations. No missing values were found, so no rows or values needed to be filled or dropped.
- Duplicate rows were checked using `.duplicated().sum()`. No duplicates were identified.
- Numeric columns were already stored as integers, so no numeric type conversion was necessary.
- Text columns were correctly stored as text.
- No date conversion was required because the dataset does not contain a date column.
- The numeric identifier columns were reviewed for unusual values. They were not treated as traditional outliers because they represent IDs rather than measurements.

## Part 2 Checkpoint Questions

### 1. How many missing values did you find? In which columns?

No missing values were found. The `userId`, `id`, `title`, and `body` columns each contained 0 missing values.

### 2. Did you drop rows, fill values, or keep them? Why?

No rows or values were dropped or filled because the dataset contained no missing values. The original records were therefore retained.

### 3. Were there duplicates? How did you handle them?

No duplicate rows were identified. The dataset was still passed through `drop_duplicates()` as a precaution.

### 4. Did any data type conversions surprise you?

No unexpected data types were found. The ID columns were stored as integers and the title and body columns were stored as text, which was appropriate for the dataset.

### PART 3 EDA

In [42]:
cleaned_df.columns

Index(['postId', 'id', 'name', 'email', 'body'], dtype='str')

In [43]:
cleaned_df["name_length"] = cleaned_df["name"].astype(str).str.len()

In [44]:
cleaned_df["name_length"].describe()

count    500.000000
mean      40.552000
std       14.968079
min       10.000000
25%       29.000000
50%       40.000000
75%       52.000000
max       81.000000
Name: name_length, dtype: float64

In [45]:
#Checking body lengths
cleaned_df["body_length"] = cleaned_df["body"].astype(str).str.len()

In [46]:
cleaned_df["body_length"].describe()

count    500.000000
mean     161.528000
std       29.392807
min       78.000000
25%      140.750000
50%      159.000000
75%      180.000000
max      263.000000
Name: body_length, dtype: float64

In [47]:
#Checking for skewness
cleaned_df[["name_length", "body_length"]].skew()

name_length    0.157392
body_length    0.244217
dtype: float64

In [51]:
cleaned_df[["id","name_length", "body_length"]].corr()

,id,name_length,body_length
id,1.000000,0.017231,0.103202
name_length,0.017231,1.000000,-0.063774
body_length,0.103202,-0.063774,1.000000


In [52]:
#Finding strongest correlation between name_length and body_length and id
correlation_matrix = cleaned_df[["id","name_length", "body_length"]].corr()

In [53]:
correlation_matrix.abs()

,id,name_length,body_length
id,1.000000,0.017231,0.103202
name_length,0.017231,1.000000,0.063774
body_length,0.103202,0.063774,1.000000


In [57]:
cleaned_df.groupby("id")["body_length"].mean()
average_body_length = cleaned_df.groupby("id")["body_length"].mean()

In [58]:
average_body_length.sort_values(ascending=False)

id
22     263.0
374    247.0
140    244.0
278    243.0
366    239.0
       ...  
354    101.0
176     98.0
346     96.0
179     94.0
452     78.0
Name: body_length, Length: 500, dtype: float64

In [59]:
cleaned_df["id"].value_counts()

id
1      1
2      1
3      1
4      1
5      1
      ..
496    1
497    1
498    1
499    1
500    1
Name: count, Length: 500, dtype: int64

In [60]:
cleaned_df["id"].mode()

0        1
1        2
2        3
3        4
4        5
      ... 
495    496
496    497
497    498
498    499
499    500
Name: id, Length: 500, dtype: int64

In [61]:
average_body_length.idxmax()

np.int64(22)

In [63]:
average_body_length.max()

np.float64(263.0)

In [66]:
#Look for unusual records
cleaned_df.loc[cleaned_df["body_length"] == cleaned_df["body_length"].max()]

,postId,id,name,email,body,name_length,body_length
21,5,22,porro repellendus aut tempore quis hic,Khalil@emile.co.uk,qui ipsa animi nostrum praesentium voluptatibu...,38,263


In [67]:
cleaned_df.loc[cleaned_df["body_length"] == cleaned_df["body_length"].min()]

,postId,id,name,email,body,name_length,body_length
451,91,452,quia excepturi in harum repellat consequuntur ...,General@schuyler.org,ratione sequi sed\nearum nam aut sunt\nquam cu...,57,78


## EDA Findings

The numerical analysis showed differences in the lengths of post titles and post bodies. The `body_length` variable showed more variation than the title length.

The correlation analysis was used to identify relationships between user IDs, post IDs, title length, and body length. Identifier variables such as `id` should be interpreted carefully because they do not represent meaningful continuous measurements.

Grouping posts by `userId` showed differences in average post body length between users.

An interesting pattern was that some posts had noticeably longer bodies than others. These records were investigated as potential unusual observations, but they were not removed because a long post is not necessarily an error.

## Part 3 Checkpoint Questions

### 1. Which column has the most missing data?

All columns had 0 missing values, so there was no single column with more missing data than the others.

### 2. What is the correlation between your strongest pair of variables?

The strongest correlation was between `[id]` and `[body_length]`, with a correlation coefficient of `[0.103202]`.

### 3. What's the most common value in your categorical column?

The most common value in the selected categorical column was `[VALUE]`.

### 4. What's one unexpected pattern you found?

One interesting pattern was `[the data set has limited missing values]`. This was identified through the correlation and groupby analysis.

### CONCLUSION
# Conclusion

The dataset was successfully loaded, inspected, cleaned, and analyzed using Pandas.

The cleaning process showed that the dataset contained no missing values or duplicate rows, and the existing data types were appropriate. No records were removed as outliers because the numerical identifier columns were not suitable for traditional outlier analysis.

The exploratory analysis provided information about the distribution of numerical variables, categorical values, correlations, and differences between users.

A useful next step would be to create visualizations such as histograms, bar charts, and scatter plots to communicate the patterns identified during the EDA more clearly. Further analysis could also investigate relationships between post length and user activity.